MekHQ/src/mekhq/campaign/mission/ScenarioObjectiveProcessor.java

```diff
@@ -174,15 +174,15 @@ public void updateObjectiveEntityState(Entity entity, boolean forceEntityEscape,
                         break;
                     case PreventReachMapEdge:
                         entityMeetsObjective = forceEntityDestruction ||
-                            !forceEntityEscape && !entityHasReachedDestinationEdge(entity, objective, opponentHasBattlefieldControl);
+                                !entityHasReachedDestinationEdge(entity, objective);
```

Comment:
The logic here was originally that the unit automatically counts as having reached the edge if the opposing force retreats. So if the player answers "no" to "did you have battlefield control", then this objective should automatically fail.

Reply:
I understand that and I am saying that this is bad logic. See the example in my OP. This is a fairly common example of what a breakthrough scenario might look like - hold off a numerically superior force for some time before retreating. Note an equally important case is also made impossible on the other side when the player is the breakthrough force - trying to break through a numerically superior force. In both cases, the player would not have battlefield control. 

My basic point is that the battlefield control bit basically assumes too much about the context of the scenario and as a result it reduces the flexibility of this objective. I think that is a poor design decision. Removing it simply makes the objective literally work as intended - the goal is to either exit your units or prevent the exit of other units to a certain map edge. 

This logic also make these two objectives largely superfluous. If you lose the objective because you do not have battlefield control, then in practice this objective is about establishing battlefield control, which makes it like most other objectives and diminishes the ability to give players a more diverse set of potential objectives.

Reply:
Sounds like there needs to be two types of "Reach Edge" objective then - one affected by battlefield control and one not.

Reply:
Maybe, but my guess is that the other type could probably be accomplished by existing scenario objectives. In this case, I think the Destroy objective in combination with a bot whose behavior is to exit via the player's home map edge and forced withdrawal set to false would basically accomplish this. 

In regards to tradeoffs, @NickAragua, for the PreventBreakthrough objective there is no tradeoff option. If you do not control the battlefield, then you will lose unless you destroy or cripple enough of the opposing force. Again, if that is the goal, it can be handled by other objectives.

Reply:
I'm thinking of the following scenario:

You have bot mechs trying to cross over to the east side of a battlefield. If the player gives up control of the battlefield (by retreating or being annihilated), then any hostile mechs still on the field (barring immobilized and possibly crippled ones) should count has having reached the east side, even if they haven't done so at the point when the player ended the scenario. 

Unless I'm misreading the logic, this pull request makes it a requirement that the bot units *actually* cross the east edge even if the player's force has already been annihilated.

Reply:
Well, thats exactly the problem. Whether the hostile mechs still on the field should count as having broken through depends on the context of the scenario. In my opening example, it should not if the player held out for the requisite number of turns before ending the scenario. In a scenario without a time limit, it s reasonable to assume those mechs will breakthrough. 

I think the missing feature that would allow us to distinguish the cases is really taking account of the time limits. Thats something I would like to do eventually, as I would like to improve scenario objectives in the long run, but that is a much bigger project. 

In the interim, I do not think the scenario you outline should be a reason to leave the logic. In some cases, until we work out a more complete system, players may have to correct things in the resolution wizard. But this scenario should be relatively rare as the bot's goal is not to destroy player units but to reach the other side. So, they are not sticking around to finish off heavily damaged units. And in situations where this occurs, the player can still get the correct outcome by continuing the game until the bot fully exits the map, which it will continue to do. Similarly in a breakthrough, a player who completely destroys the opposing force can continue the game and exit all of their units and I would argue that is precisely what they should do.

Reply:
I should also note that if you are running a prevent breakthrough scenario without a time limit, then you are really better off to set the scenario objective as destroy rather than prevent breakthrough because this is actually what the player is being asked to do - destroy the units before they exit. In other words, I think the issue you are concerned with @NickAragua can mostly be handled by the existing architecture. It just requires some rethinking about how scenario objectives are assigned.

Reply:
I don't think we're going to come to an agreement on this, but one other point is that there are already several scenarios that use these objective types and rely on them working the way they work currently, with no time limit involved. So if you want to modify one of those to work roughly the way it worked previously as an example of what you're thinking, then I'm fine letting this go.

Some examples (all in data/scenariotemplates)
"Breakthrough.xml"
"Prevent Breakthrough.xml"
"Pursuit.xml"
"Harass.xml"

Reply:
Fixed almost all of these issues in commit  8d8e07121bad24. Nine templates involve either the Breakthrough or PreventBreakthrough objective but only two actually need to be changed:

* Breakthrough.xml - This works fine as is. If the player clears the board of opposition, they can walk off the map themselves and meet the objective. Note that if you don't want to make players walk off then you could also create another objective to ForceWithdraw 100% of the enemy force to win. If that is the preferred solution, I can add that code. Same applies to several similar templates below.
* Diversion.xml - This appears to be more or less the same as Breakthrough.xml with some slight tweaks to parameters, so the same logic applies. If the player clears the board, they can walk off to meet the objective.
* Pursuit.xml  - this is very similar to Breakthrough.xml but also includes logic where the player can win by breakthrough or by destroying/routing 50% of the OpFor. So this will work fine as is.
* PreventBreakthrough.xml and Harrass.xml - These are the two cases that I changed. The objective is now to destroy at least 50% of the opposing force before they reach the other map edge.  This is functionally equivalent to the prior objective but works with both old and new logic. It is literally the goal here.
* ConvoyEscort.xml - As with Pursuit.xml you can escort the units to the map edge or route the enemy force to win, so no issue here.
* Low Atmo - Escort.xml and Space - Blockade Run Escort.xml - As with Pursuit.xml, you can escort the units to the map edge or route the enemy force to win, so no issue here.
* ConvoyAttack.xml - This is the only one, I have not sorted out at the moment, because its logic is the most complicated, but I will work more on it later.

So, aside from the one remaining template which I need to fix, I think the overall impact here on existing ScenarioTemplates is actually relatively minimal and can easily be changed to actually better reflect what is being asked of the player in the two cases that need to be changed.

Reply:
All right, I think we've reached a reasonable compromise here.